server/src/main/java/org/elasticsearch/indices/recovery/RecoverySourceHandler.java

```diff
@@ -982,13 +983,19 @@ void sendFiles(Store store, StoreFileMetadata[] files, IntSupplier translogOps,
                 protected void onNewResource(StoreFileMetadata md) throws IOException {
                     offset = 0;
                     IOUtils.close(currentInput, () -> currentInput = null);
-                    final IndexInput indexInput = store.directory().openInput(md.name(), IOContext.READONCE);
-                    currentInput = new InputStreamIndexInput(indexInput, md.length()) {
-                        @Override
-                        public void close() throws IOException {
-                            IOUtils.close(indexInput, super::close); // InputStreamIndexInput's close is a noop
-                        }
-                    };
+                    // Open all files other than Segments* using IOContext.READ.
+                    // With Lucene9_12 a READONCE context will confine the underlying IndexInput (MemorySegmentIndexInput) to a single thread.
+                    // Segments* files require IOContext.READONCE
+                    // https://github.com/apache/lucene/blob/b2d3a2b37e00f19a74949097736be8fd64745f61/lucene/test-framework/src/java/org/apache/lucene/tests/store/MockDirectoryWrapper.java#L817
+                    if (md.name().startsWith(IndexFileNames.SEGMENTS) == false) {
+                        final IndexInput indexInput = store.directory().openInput(md.name(), IOContext.READ);
+                        currentInput = new InputStreamIndexInput(indexInput, md.length()) {
+                            @Override
+                            public void close() throws IOException {
+                                IOUtils.close(indexInput, super::close); // InputStreamIndexInput's close is a noop
+                            }
+                        };
+                    }
```

Comment:
Does this need an `else` clause?  As I read it this means that we don't set `currentInput` at all when the new resource is a segments file?

Reply:
Oh hang on, I see that this is done in `readBytes()`.  It feels a bit awkward to split it across two methods though?

Reply:
I've tried this: https://github.com/crate/crate/pull/16805 but it leads to multiple failures.

Reply:
Thanks, probably not worth spending any more time on then.

Reply:
thx!